### PR TITLE
Balder/pass stringref by value

### PR DESCRIPTION
**Pull request by baldersheim**
@toregge PR
@vekterli FYI

### CODE REVIEW

**Review by toregge — CHANGES_REQUESTED** [review]
Compilation fails on Fedora 28, using system compiler (gcc 8.1.1).

Otherwise, the changes looks OK.

File: document/src/vespa/document/fieldvalue/stringfieldvalue.cpp [hunk 1]
```diff
@@ -109,7 +109,7 @@ FieldValue & StringFieldValue::assign(const FieldValue & rhs)
     if (rhs.inherits(StringFieldValue::classId)) {
         *this = static_cast<const StringFieldValue &>(rhs);
     } else {
-        *this = static_cast<const stringref &>(rhs.getAsString());
+        *this = static_cast<stringref >(rhs.getAsString());
```
[the remark this answers is not in the document]
Comment (toregge) [reply document/src/vespa/document/fieldvalue/stringfieldvalue.cpp]: changing the line to

*this = rhs.getAsString().operator stringref();

seems to work.

File: searchlib/src/vespa/searchlib/features/attributefeature.cpp [hunk 2]
```diff
@@ -37,18 +37,18 @@ using namespace search::fef::indexproperties;
...
 template <typename X, typename Y>
-bool equals(const X & lhs, const Y & rhs) {
+bool equals(X lhs, Y rhs) {
     return lhs == rhs;
 }
 
 template <>
-bool equals<ConstCharPtr, vespalib::stringref>(const ConstCharPtr & lhs, const vespalib::stringref & rhs) {
+bool equals<ConstCharPtr, vespalib::stringref>(ConstCharPtr lhs, vespalib::stringref rhs) {
     return strcmp(lhs, rhs.data()) == 0;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/attributefeature.cpp]: Yes, I am aware of that, but that is nothing new here. And we require that stringattributes are zero terminated.

File: staging_vespalib/src/vespa/vespalib/util/xmlstream.cpp [hunk 3]
```diff
@@ -446,18 +446,18 @@ XmlContentWrapper::XmlContentWrapper(const char* value, uint32_t size)
...
-template XmlAttribute::XmlAttribute(const std::string &, const ConstCharP &, unsigned int);
-template XmlAttribute::XmlAttribute(const std::string &, const bool &, unsigned int);
-template XmlAttribute::XmlAttribute(const std::string &, const int16_t &, unsigned int);
-template XmlAttribute::XmlAttribute(const std::string &, const int32_t &, unsigned int);
-template XmlAttribute::XmlAttribute(const std::string &, const int64_t &, unsigned int);
-template XmlAttribute::XmlAttribute(const std::string &, const uint16_t &, unsigned int);
-template XmlAttribute::XmlAttribute(const std::string &, const uint32_t &, unsigned int);
-template XmlAttribute::XmlAttribute(const std::string &, const uint64_t &, unsigned int);
-template XmlAttribute::XmlAttribute(const std::string &, const double &, unsigned int);
+template XmlAttribute::XmlAttribute(const std::string &, std::string, unsigned int);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply staging_vespalib/src/vespa/vespalib/util/xmlstream.cpp]: I hope the xml stream dies before that. It is not used for anything, but debugging I think.

File: document/src/vespa/document/fieldvalue/stringfieldvalue.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/fieldvalue/stringfieldvalue.cpp]: Thanks, fixed.

**Review by toregge — APPROVED** [review]
LGTM

**Pull request merged**